hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -63,217 +86,221 @@ data NarEffects (m :: * -> *) = NarEffects {
+
+
+-- | This implementation of Nar encoding takes an arbitrary @yield@
+--   function from any streaming library, and repeatedly calls
+--   it while traversing the filesystem object to Nar encode
+streamNarIO
+  :: forall m n.(IO.MonadIO m, Monad n)
+  => (BS.ByteString -> m ())
+  -> NarEffects IO
+  -> FilePath
+  -> m (Nar n)
+streamNarIO yield effs basePath = do
+  yield (str "nix-archive-1")
+  fmap Nar $ parens (go basePath)
+  where
 
--- | A FileSystemObject (FSO) is an anonymous entity that can be NAR archived
-data FileSystemObject =
-    Regular IsExecutable Int64 BSL.ByteString
-    -- ^ Reguar file, with its executable state, size (bytes) and contents
-  | Directory (Map.Map FilePathPart FileSystemObject)
-    -- ^ Directory with mapping of filenames to sub-FSOs
-  | SymLink T.Text
-    -- ^ Symbolic link target
-  deriving (Eq, Show)
+    go :: FilePath -> m (FileSystemObject n)
+    go path = do
+      isDir     <- IO.liftIO $ narIsDir effs path
+      isSymLink <- IO.liftIO $ narIsSymLink effs path
+      let isRegular = not (isDir || isSymLink)
+      case isSymLink of
+
+        True -> do
+          -- Symlink
+          target <- IO.liftIO $ narReadLink effs path
+          yield $
+            strs ["type", "symlink", "target", BSC.pack target]
+          return $ SymLink target
+
+        False -> case isDir of
+          False -> do
+            -- Regular File
+            isExec <- IO.liftIO $ isExecutable effs path
+            yield $ strs ["type","regular"]
+            when (isExec == Executable) (yield $ strs ["executable", ""])
+            fSize <- IO.liftIO $ narFileSize effs path
+            yield $ str "contents"
+            yield $ int fSize
+            yieldFile path fSize
+            return $ Regular (\(e :: NarEffects n) ->
+                                (,,) <$> isExecutable e path
+                                     <*> narFileSize e path
+                                     <*> narReadFile e path
+                             )
+          True -> do
+            -- Directory
+            fs <- IO.liftIO (narListDir effs path)
+            yield $ strs ["type", "directory"]
+            fsos <- forM (List.sort fs) $ \f -> do
+              yield $ str "entry"
+              fso <- parens $ do
+                let fullName = path </> f
+                yield (strs ["name", BSC.pack f, "node"])
+                fso <- parens (go fullName)
+                return fso
+              return (f,fso)
+            return $ Directory $ Map.fromList fsos
+
+    str :: BS.ByteString -> BS.ByteString
+    str t = let len =  BS.length t
+            in  int len <> padBS len t
+
+    padBS :: Int -> BS.ByteString -> BS.ByteString
+    padBS strSize bs = bs <> BS.replicate (padLen strSize) 0
+
+    parens act = do
+      yield (str "(")
+      r <- act
+      yield (str ")")
+      return r
+
+    -- Read, yield, and pad the file
+    -- yieldFile :: FilePath -> Int64 -> Streams.Generator BS.ByteString ()
```

Comment:
TIL you can do ``yieldField :: _`` and get 
```
Found type wildcard ‘_’ standing for ‘FilePath -> Int64 -> m ()’
```

Reply:
Thanks!